java.lang.NoClassDefFoundError: com/oracle/svm/core/jdk/LocalizationFeature

## Describe the bug
When running `integration-tests/jpa-mariadb` on quarkus master with a recent [Graal VM CE build](https://github.com/graalvm/graalvm-ce-dev-builds/releases/tag/21.1.0-dev-20210323_1605) (21.1.0-dev) in `-Dnative` mode I see NCDFE and ClassNotFoundException for Graal VMs `LocalizationFeature`. This seems to be related to Quarkus' generated autofeature.

### Expected behavior
No stacktrace when generating a native image for `integration-tests/jpa-mariadb`

### Actual behavior
```
[quarkus-integration-test-jpa-mariadb-999-SNAPSHOT-runner:522013]    classlist:   7,816.03 ms,  1.19 GB
[quarkus-integration-test-jpa-mariadb-999-SNAPSHOT-runner:522013]        (cap):   1,057.90 ms,  1.19 GB
[quarkus-integration-test-jpa-mariadb-999-SNAPSHOT-runner:522013]        setup:   5,911.44 ms,  1.19 GB
java.lang.NoClassDefFoundError: com/oracle/svm/core/jdk/LocalizationFeature
	at io.quarkus.runner.AutoFeature.beforeAnalysis(AutoFeature.zig:1030)
	at com.oracle.svm.hosted.NativeImageGenerator.lambda$runPointsToAnalysis$13(NativeImageGenerator.java:724)
	at com.oracle.svm.hosted.FeatureHandler.forEachFeature(FeatureHandler.java:71)
	at com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:724)
	at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:580)
	at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$2(NativeImageGenerator.java:493)
	at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1407)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Caused by: java.lang.ClassNotFoundException: com.oracle.svm.core.jdk.LocalizationFeature
	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:471)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:589)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	... 12 more
17:57:19,926 INFO  [org.hib.Version] HHH000412: Hibernate ORM core version 5.4.30.Final
17:57:19,948 INFO  [org.hib.ann.com.Version] HCANN000001: Hibernate Commons Annotations {5.1.2.Final}
17:57:20,036 INFO  [org.hib.dia.Dialect] HHH000400: Using dialect: org.hibernate.dialect.MariaDB103Dialect
```


## To Reproduce
Steps to reproduce the behavior:
1.  Build quarkus master at revision d185939eb2e1ba22b817c58ffc51865a0e66669d
2.  Download a GraalVM CE dev build from https://github.com/graalvm/graalvm-ce-dev-builds/releases/tag/21.1.0-dev-20210323_1605
3.  Set GRAALVM_HOME to the downloaded Graal VM CE build
4.  `$ ./mvnw verify -Dnative -pl integration-tests/jpa-mariadb/`

## Environment (please complete the following information):
Linux, Java 11.

### GraalVM version (if different from Java)
21.1.0-dev

### Quarkus version or git rev
d185939eb2e1ba22b817c58ffc51865a0e66669d

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)
```
$ ./mvnw --version
Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
Maven home: /home/sgehwolf/.m2/wrapper/dists/apache-maven-3.6.3-bin/1iopthnavndlasol9gbrbg6bf2/apache-maven-3.6.3
Java version: 11.0.10, vendor: Red Hat, Inc., runtime: /usr/lib/jvm/java-11-openjdk-11.0.10.0.9-1.fc33.x86_64
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "5.11.7-200.fc33.x86_64", arch: "amd64", family: "unix"
```

## Additional context
This seems to be the cause of some remaining problems for the graal master quarkus workflow. See https://github.com/oracle/graal/issues/3272

--- a/core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageAutoFeatureStep.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageAutoFeatureStep.java
@@ -35,6 +35,7 @@ import io.quarkus.deployment.builditem.nativeimage.RuntimeInitializedPackageBuil
 import io.quarkus.deployment.builditem.nativeimage.RuntimeReinitializedClassBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.ServiceProviderBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.UnsafeAccessedFieldBuildItem;
+import io.quarkus.gizmo.AssignableResultHandle;
 import io.quarkus.gizmo.CatchBlockCreator;
 import io.quarkus.gizmo.ClassCreator;
 import io.quarkus.gizmo.ClassOutput;
@@ -66,7 +67,8 @@ public class NativeImageAutoFeatureStep {
     static final String JNI_RUNTIME_ACCESS = "com.oracle.svm.core.jni.JNIRuntimeAccess";
     static final String BEFORE_ANALYSIS_ACCESS = Feature.BeforeAnalysisAccess.class.getName();
     static final String DYNAMIC_PROXY_REGISTRY = "com.oracle.svm.core.jdk.proxy.DynamicProxyRegistry";
-    static final String LOCALIZATION_FEATURE = "com.oracle.svm.core.jdk.LocalizationFeature";
+    static final String LEGACY_LOCALIZATION_FEATURE = "com.oracle.svm.core.jdk.LocalizationFeature";
+    static final String LOCALIZATION_FEATURE = "com.oracle.svm.core.jdk.localization.LocalizationFeature";
 
     @BuildStep
     void generateFeature(BuildProducer<GeneratedNativeImageClassBuildItem> nativeImageClass,
@@ -217,12 +219,28 @@ public class NativeImageAutoFeatureStep {
         }
 
         if (!resourceBundles.isEmpty()) {
-            ResultHandle locClass = overallCatch.loadClass(LOCALIZATION_FEATURE);
+            AssignableResultHandle registerMethod = overallCatch.createVariable(Method.class);
+            AssignableResultHandle locClass = overallCatch.createVariable(Class.class);
+            TryBlock locTryBlock = overallCatch.tryBlock();
+            ResultHandle legacyLocClass = locTryBlock.loadClass(LEGACY_LOCALIZATION_FEATURE);
+            locTryBlock.assign(locClass, legacyLocClass);
 
-            ResultHandle params = overallCatch.marshalAsArray(Class.class, overallCatch.loadClass(String.class));
-            ResultHandle registerMethod = overallCatch.invokeVirtualMethod(
-                    ofMethod(Class.class, "getDeclaredMethod", Method.class, String.class, Class[].class), locClass,
-                    overallCatch.load("addBundleToCache"), params);
+            ResultHandle legacyParams = locTryBlock.marshalAsArray(Class.class, locTryBlock.loadClass(String.class));
+            ResultHandle legacyRegisterMethod = locTryBlock.invokeVirtualMethod(
+                    ofMethod(Class.class, "getDeclaredMethod", Method.class, String.class, Class[].class), legacyLocClass,
+                    locTryBlock.load("addBundleToCache"), legacyParams);
+            locTryBlock.assign(registerMethod, legacyRegisterMethod);
+
+            CatchBlockCreator locCatchBlock = locTryBlock.addCatch(NoClassDefFoundError.class);
+            ResultHandle newLocClass = locCatchBlock.loadClass(LOCALIZATION_FEATURE);
+            locCatchBlock.assign(locClass, newLocClass);
+
+            ResultHandle newParams = locCatchBlock.marshalAsArray(Class.class, locCatchBlock.loadClass(String.class));
+            ResultHandle newRegisterMethod = locCatchBlock.invokeVirtualMethod(
+                    ofMethod(Class.class, "getDeclaredMethod", Method.class, String.class, Class[].class), newLocClass,
+                    locCatchBlock.load("prepareBundle"), newParams);
+            locCatchBlock.assign(registerMethod, newRegisterMethod);
+
             overallCatch.invokeVirtualMethod(ofMethod(AccessibleObject.class, "setAccessible", void.class, boolean.class),
                     registerMethod, overallCatch.load(true));
 

